services/QuillLMS/engines/evidence/spec/models/evidence/research/gen_ai/llm_prompt_spec.rb

```diff
@@ -42,12 +42,60 @@ module GenAI
         it { should have_readonly_attribute(:suboptimal_examples_count) }
 
         it { should have_one(:trial).dependent(:destroy) }
+        it { should have_one(:llm).through(:trial) }
         it { should have_many(:llm_prompt_prompt_examples).dependent(:destroy) }
         it { should have_many(:prompt_examples).through(:llm_prompt_prompt_examples) }
         it { should have_many(:llm_prompt_guidelines).dependent(:destroy) }
         it { should have_many(:guidelines).through(:llm_prompt_guidelines) }
+        it { should have_many(:llm_prompt_relevant_texts).dependent(:destroy) }
+        it { should have_many(:relevant_texts).through(:llm_prompt_relevant_texts) }
 
         it { should have_many(:activated_llm_prompts).class_name('Evidence::ActivatedLLMPrompt').dependent(:destroy) }
+
+        describe '.create_from_template!' do
+          subject do
+            described_class.create_from_template!(
+              dataset_id:,
+              guideline_ids:,
+              llm_prompt_template_id:,
+              prompt_example_ids:,
+              relevant_text_id:
+            )
+          end
+
+          let(:stem_vault) { create(:evidence_research_gen_ai_stem_vault) }
+          let(:dataset) { create(:evidence_research_gen_ai_dataset, stem_vault:) }
+          let(:dataset_id) { dataset.id }
+          let(:guidelines) { create_list(:evidence_research_gen_ai_guideline, 2, dataset:) }
+          let(:guideline_ids) { guidelines.map(&:id) }
+          let(:llm_prompt_template) { create(:evidence_research_gen_ai_llm_prompt_template) }
+          let(:llm_prompt_template_id) { llm_prompt_template.id }
+          let(:prompt_examples) { create_list(:evidence_research_gen_ai_prompt_example, 2, dataset:) }
+          let(:prompt_example_ids) { prompt_examples.map(&:id) }
+
+          before { allow(LLMPromptBuilder).to receive(:run).and_return('Generated prompt') }
+
+          context 'with relevant_text_id' do
+            let(:relevant_text) { create(:evidence_research_gen_ai_relevant_text, stem_vault:) }
+            let(:relevant_text_id) { relevant_text.id }
+
+            before { create(:evidence_research_gen_ai_dataset_relevant_text, dataset:, relevant_text:) }
+
+            it { expect(subject).to be_persisted }
```

Comment:
Today I discovered that RSpec will use metaprogramming to predicate-match `be_X` expectations.  Neat!

Reply:
RSpec is a really cool demonstration of metaprogramming, also check out other constructs like `let`, `subject`, and `before`.  If you want to go down a rabbit hole, look through the internals of FactoryBot.

Now that you mention, it feels like most effective ruby DSLs end up leveraging metaprogramming.